rq/group.py

```diff
@@ -0,0 +1,124 @@
+from typing import List, Optional
+from uuid import uuid4
+
+from redis import Redis
+from redis.client import Pipeline
+
+from . import Queue
+from .exceptions import NoSuchGroupError
+from .job import Job
+from .queue import EnqueueData
+from .utils import as_text
+
+
+class Group:
+    """A Group is a container for tracking multiple jobs with a single identifier."""
+
+    REDIS_GROUP_NAME_PREFIX = 'rq:group:'
+    REDIS_GROUP_KEY = 'rq:groups'
+
+    def __init__(self, connection: Redis, id: str = None):
+        self.id = id if id else str(uuid4())
+        self.connection = connection
+        self.key = '{0}{1}'.format(self.REDIS_GROUP_NAME_PREFIX, self.id)
+
+    def __repr__(self):
+        return "Group(id={})".format(self.id)
+
+    def _add_jobs(self, jobs: List[Job], pipeline: Optional['Pipeline'] = None):
+        """Add jobs to the group"""
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        pipe.sadd(self.key, *[job.id for job in jobs])
+        pipe.sadd(self.REDIS_GROUP_KEY, self.id)
+        if pipeline is None:
+            pipe.execute()
+
+    def cleanup(self, pipeline: Optional['Pipeline'] = None):
+        """Delete jobs from the group's job registry that have been deleted or expired from Redis.
+        We assume while running this that alive jobs have all been fetched from Redis in fetch_jobs method"""
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        Group.cleanup_group(self.id, self.connection, pipeline=pipe)
+        if pipeline is None:
+            pipe.execute()
+
+    def enqueue_many(self, queue: Queue, job_datas: List['EnqueueData'], pipeline: Optional['Pipeline'] = None):
+        pipe = pipeline if pipeline else self.connection.pipeline()
+
+        jobs = queue.enqueue_many(job_datas, group_id=self.id, pipeline=pipe)
+
+        self._add_jobs(jobs, pipeline=pipe)
+
+        if pipeline is None:
+            pipe.execute()
+
+        return jobs
+
+    def get_jobs(self) -> list:
+        """Retrieve list of job IDs from the group key in Redis"""
+        job_ids = [as_text(job) for job in self.connection.smembers(self.key)]
+        return [job for job in Job.fetch_many(job_ids, self.connection) if job is not None]
+
+    def delete(self):
+        self.connection.delete(self.key)
+
+    def delete_job(self, job_id: str, pipeline: Optional['Pipeline'] = None):
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        pipe.srem(self.key, job_id)
+        if pipeline is None:
+            pipe.execute()
+
+    @classmethod
+    def create(cls, connection: Redis, id: Optional[str] = None):
+        return cls(id=id, connection=connection)
+
+    @classmethod
+    def fetch(cls, id: str, connection: Redis):
+        """Fetch an existing group from Redis"""
+        group = cls(id=id, connection=connection)
+        group.cleanup()
```

Comment:
I think we shouldn't call `cleanup()` everytime a group is fetched but rather only when job IDs are read from it.

Reply:
Done